presto-hive/src/main/java/io/prestosql/plugin/hive/util/FieldSetterFactory.java

```diff
@@ -361,24 +375,60 @@ public void setField(Block block, int position)
         }
     }
 
-    private static class TimestampFieldSetter
+    private abstract static class TimestampFieldSetter
             extends FieldSetter
     {
-        private final DateTimeZone timeZone;
-        private final TimestampWritableV2 value = new TimestampWritableV2();
+        protected final DateTimeZone timeZone;
+        protected final TimestampWritableV2 value = new TimestampWritableV2();
 
         public TimestampFieldSetter(SettableStructObjectInspector rowInspector, Object row, StructField field, DateTimeZone timeZone)
         {
             super(rowInspector, row, field);
             this.timeZone = requireNonNull(timeZone, "timeZone is null");
         }
 
+        protected Timestamp getTimestamp(long micros, IntSupplier picosSupplier)
+        {
+            long epochMillis = floorDiv(micros, MICROSECONDS_PER_MILLISECOND);
+            epochMillis = timeZone.convertLocalToUTC(epochMillis, false);
+            int nanosFromMicros = floorMod(micros, MICROSECONDS_PER_SECOND) * NANOSECONDS_PER_MICROSECOND;
+            int nanosFromPicos = (int) roundDiv(picosSupplier.getAsInt(), PICOSECONDS_PER_NANOSECOND);
```

Comment:
nit: we could add `int roundDiv(int,long)` overload for cases iike this one

you can take https://github.com/findepi/presto/commits/round-div is you want to